aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +26,236 @@ namespace at {
-  if(tensor3_counter != NULL)
-    delete [] tensor3_counter;
+  oss << "and " << tensors[tensors.size() - 1].numel()
+      << " elements respectively";
+  return oss.str();
 }
 
-/*
-  Apply a pointwise operator to three tensors.
+inline bool _apply_preamble(ArrayRef<Tensor> tensors) {
+  checkBackend("CPU_tensor_apply", tensors, Backend::CPU);
+  if (!_all_equal_numel(tensors))
+    throw std::runtime_error(_all_equal_numel_error(tensors));
+  // An empty tensor has no elements
+  for (auto& t : tensors)
+    if (t.sizes().equals({0}))
+      return false;
+  return true;
+}
 
-  The calling convention for op is a function/functor that takes takes three references to
-  type scalar; at least one of these references should be non-const in order to write the output.
-  For example, to compute a = b + c, op would be of the form:
-  [](scalar &a_val, const scalar &b_val, const scalar &c_val) { a_val = b_val + c_val; };
-*/
-template<typename scalar1, typename scalar2, typename scalar3, typename Op>
-void CPU_tensor_apply3(Tensor tensor1, Tensor tensor2, Tensor tensor3, Op op) {
-  CPU_tensor_apply3_dim<scalar1, scalar2, scalar3, Op>(tensor1, tensor2, tensor3, -1, op);
+inline void iterate_all(){};
+
+template <typename Arg, typename... Args>
+inline void iterate_all(Arg& iter, Args&... iter_tail) {
+  iterate(iter);
+  iterate_all(iter_tail...);
 }
 
-template <typename scalar1, typename scalar2, typename scalar3, typename scalar4, typename Op>
-void CPU_tensor_apply4_dim(Tensor &tensor1, Tensor& tensor2, Tensor& tensor3, Tensor& tensor4, int64_t dim, Op op) {
-  checkBackend("CPU_tensor_apply4", {tensor1, tensor2, tensor3, tensor4}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar3, tensor3, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar4, tensor4, dim, 1)
+inline bool continue_loop() {
+  return true;
+}
 
-  int elements_equal = 1;
-  auto t1_numel = tensor1.numel();
-  auto t2_numel = tensor2.numel();
-  auto t3_numel = tensor3.numel();
-  auto t4_numel = tensor4.numel();
-  if(t1_numel!= t2_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t3_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t4_numel) {
-      elements_equal = 0;
-  }
-  if (elements_equal == 0) {
-    std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << ", " << tensor2.sizes() << ", "
-        << tensor3.sizes() << ", and " << tensor4.sizes() << " to have the same number of elements, but got "
-        << t1_numel << ", " << t2_numel << ", " << t3_numel << ", and " << t4_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
+template <typename Arg, typename... Args>
+inline bool continue_loop(Arg& iter, Args&... iter_tail) {
+  bool result = false;
+  if (iter.dim_ != 0) {
```

Comment:
Is it just me or would `iter.dim_ == 0` mean that the loop will never terminate?

Reply:
I think this function will return "false" so the loop won't be run anymore, since result is false by default and it returns result && continue_loop(...).

Reply:
Uh, nvm... It's late. But still, when would that be the case? Wouldn't we break out of apply in the preamble if those tensors wouldn't have any elems?

Reply:
0-dim has 1 element, so we still need to apply the op. It's annoying and I can't find a way to treat it cleanly.

Reply:
Can't we just special case 0-dim tensors to break out before the loop, and then avoid all those checks at every iteration?

Reply:
It's worth trying.